pool/heap.go

```diff
@@ -0,0 +1,223 @@
+	)
+
+	sort.Sort(BucketByCapacity(b))
+
+	h := heap{m: heapMetrics{
+		overflows: m.Counter("overflows"),
+		misplaces: m.Counter("misplaces"),
+	}}
+
+	for _, cfg := range b {
+		m := m.Tagged(map[string]string{
+			"bucket-capacity": strconv.Itoa(cfg.Capacity),
+		})
+
+		s := &slot{class: cfg.Capacity, opts: NativePoolOptions{
+			Size: uint(cfg.Count),
+			Type: reflect.ArrayOf(cfg.Capacity, ByteType),
+		}, m: slotMetrics{
+			used: m.Gauge("used"),
+			size: m.Gauge("size"),
+		}}
+
+		h.slots = append(h.slots, s)
+	}
+
+	return h
+}
+
+type heap struct {
+	slots []*slot
+
+	m heapMetrics
+}
+
+type heapMetrics struct {
+	overflows tally.Counter
+	misplaces tally.Counter
+}
+
+type slot struct {
+	sync.RWMutex
+
+	class int
+	opts  NativePoolOptions
+	pools []NativePool
+
+	m slotMetrics
+}
+
+type slotMetrics struct {
+	used tally.Gauge
+	size tally.Gauge
+}
+
+func (s *slot) get() interface{} {
+	if segment := s.getOr(s.RLocker(), func() interface{} {
+		return nil
+	}); segment != nil {
+		return segment
+	}
+
+	// Slow path - double-check that there are no segments left,
+	// then grow while holding an exclusive lock.
+	return s.getOr(s, func() interface{} {
+		s.pools = append(
+			[]NativePool{NewNativePool(s.opts)}, s.pools...)
+		return s.pools[0].Get()
+	})
+}
+
+func (s *slot) getOr(l sync.Locker, fn OverflowFn) interface{} {
+	var segment interface{}
+
+	l.Lock()
+
+	for _, pool := range s.pools {
+		if segment = pool.GetOr(func() interface{} {
+			return nil
+		}); segment != nil {
+			l.Unlock()
+			s.updateMetrics()
+			return segment
+		}
+	}
+
+	segment = fn()
+
+	l.Unlock()
+
+	return segment
+}
+
+func (s *slot) put(segment interface{}) {
+	s.RLock()
+
+	for i := range s.pools {
+		if s.pools[i].Owns(segment) {
+			s.pools[i].Put(segment)
+			break
+		}
+	}
+
+	s.RUnlock()
+}
+
+func (s *slot) init() {
+	// Preallocate a pool for each sizeclass to save time later.
+	s.pools = []NativePool{NewNativePool(s.opts)}
+}
+
+func (s *slot) updateMetrics() {
+	// TODO(@kobolog): Use Dice.
+	if time.Now().UnixNano()%sampleObjectPoolLengthEvery != 0 {
+		return
+	}
+
+	var used, size int64
+
+	s.RLock()
+
+	for i := range s.pools {
+		a, b := s.pools[i].Size()
+
+		used += int64(a)
+		size += int64(b)
+	}
+
+	s.RUnlock()
+
+	s.m.used.Update(used)
+	s.m.size.Update(size)
+}
+
+func (p heap) Init() {
+	for _, s := range p.slots {
+		s.init()
+	}
+}
+
+func (p heap) pick(class int, action func(*slot)) bool {
+	for _, slot := range p.slots {
+		if class <= slot.class {
+			action(slot)
+			return true
+		}
+	}
+
+	return false
+}
+
+func (p heap) Get(n int) []byte {
+	var head []byte
+
+	if p.pick(n, func(s *slot) {
+		head = *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
+			Data: reflect.ValueOf(s.get()).Pointer(),
+			Len:  0,
+			Cap:  s.class}))
+	}) {
+		return head
+	}
+
+	p.m.overflows.Inc(1)
+
+	// Allocate a segment directly from the system heap.
+	return mmap(n)[:0:n]
+}
+
+func (p heap) Put(head []byte) {
```

Comment:
If it matches one of the slot sizes, no pools there will pass the `pool.Owns()` guard in `slot.put()`, and if it's big enough to end up being `munmap()`-ed, then it will simply fail because `munmap()` keeps track of mapped regions and it won't be one of them.